### PR TITLE
Mpi python cleanup

### PR DESCRIPTION
**Pull request by philbucher**
This PR cleans and refactors the gather  methods of the interface => using std vector so no more manual memory allocation
I made use of pybind-feature that allows automatic conversion btw py::list and std::vector

tests were added in the previous PR

More features will be added in a separate PR

### CODE REVIEW

**Review by jcotela — DISMISSED** [review]
Approving, good cleanup. However, please keep in mind that the functionality in this class will probably be moved/adapted to somewhere else as we advance with the MPI core.

**Comment by philbucher** [comment]
no problem @jcotela , I don't mind adapting 
Good thing there are tests for it :)

I will wait with merging for feedback from @msandre , in case you have more to add

File: external_libraries/mpi_python/mpi_python.h [hunk 1]
```diff
@@ -1,12 +1,27 @@
...
-#include <stdio.h>
-#include <iostream>
+// External includes
 
-#include "mpi.h"
-#include <pybind11/pybind11.h>
+// Project includes
 
+#ifndef KRATOS_MPI_PYTHON_H
+#define KRATOS_MPI_PYTHON_H
```
Comment (msandre) [inline external_libraries/mpi_python/mpi_python.h]: Just curious, why did you move the header guard here?

**Review by msandre — DISMISSED** [review]
Nice, much cleaner. One thing I noticed is the formatting of some of the new code has bad alignment. Could you clean this?

**Comment by philbucher** [comment]
@msandre I cleaned the formatting

File: external_libraries/mpi_python/mpi_python.h (hunk 1 above)
Comment (philbucher) [reply external_libraries/mpi_python/mpi_python.h]: I added the default header in those files (was not there before) therefore it shows up as changed

Comment (RiccardoRossi) [reply external_libraries/mpi_python/mpi_python.h]: I think @msandre  is asking why the header guard does not go at the beginning of the file, before the includes.

(i also wonder the same)

File: external_libraries/mpi_python/python_interface.cpp [hunk 2]
```diff
@@ -23,20 +38,27 @@ PYBIND11_MODULE(mpipython, m)
     RankFuncType FRank = &PythonMPI::rank;
     SizeFuncType FSize = &PythonMPI::size;
 
-    class_<PythonMPI>(m,"PythonMPI") 
+    // note that for the functions returning a vector the conversion to a python-list is automatically
+    // done by pybind, see https://github.com/pybind/pybind11/blob/master/docs/advanced/cast/stl.rst
+    std::vector<int> (PythonMPI::*gather_int)(PythonMPIComm&, const int, const int) = &PythonMPI::gather;
```
Comment (RiccardoRossi) [inline external_libraries/mpi_python/python_interface.cpp]: why do you use this mechanism here? it looks to me that the old way was actually cleaner...

i am speaking for example of 

.def("gather",&PythonMPI::gather<double>)

Comment (RiccardoRossi) [reply external_libraries/mpi_python/python_interface.cpp]: in any case i think that making the conversion explicit could be actually nicer.

for example one could do in python

    my_vec = gather....
    my_list = list( my_vec )
    my_set = set( my_vec )

i think that all of this would work automatically in python and it goes in the same direction as #2542

File: external_libraries/mpi_python/mpi_python.h (hunk 1 above)
Comment (philbucher) [reply external_libraries/mpi_python/mpi_python.h]: Ah now I understand, thx for clarifying

Will change it

File: external_libraries/mpi_python/python_interface.cpp (hunk 2 above)
Comment (philbucher) [reply external_libraries/mpi_python/python_interface.cpp]: I see, I will try to change it

The thing regarding the conversion, I am not sure if pybind allows to return an std::vector in python, the conversion to the list is done automatically

Otherwise I would have to return a list again, which imo is the uglier solution

Comment (philbucher) [reply external_libraries/mpi_python/python_interface.cpp]: Just tried both things:
- regarding the function pointers I use for exposing in python: I need those bcs the template argument is the same in both cases, therefore I cannot use it directly
- regarding the return type: Pybind throws the following error if I try to give back the vector directly to python:
~~~
TypeError: Unable to convert function return value to a Python type! The signature was
	(self: mpipython.PythonMPI, arg0: mpipython.PythonMPIComm, arg1: int, arg2: int) -> std::vector<int, std::allocator<int> >
~~~

Comment (philbucher) [reply external_libraries/mpi_python/python_interface.cpp]: @RiccardoRossi can I go ahead then?

I would like to add some more functions/capabilities in a next PR

Comment (RiccardoRossi) [reply external_libraries/mpi_python/python_interface.cpp]: Hi @philbucher 
      i think that the problem is that you didn't expose the type std::vector< std::vector< ... > > to python. 

if you give back a py::list of std::vectors, does that work?

sorry for insisting but i think that the ability to transform from/to a list or set is very important, and i really would not like to lose that ability. for example with the current version can we iterate over the return type?

something like

     for item in returned_vector_of_vectors:
          for value in item:
               print(item)

?

File: external_libraries/mpi_python/python_interface.cpp [hunk 3]
```diff
@@ -23,20 +38,27 @@ PYBIND11_MODULE(mpipython, m)
...
+
+    py::class_<PythonMPI>(m,"PythonMPI")
     .def_property_readonly("rank",FRank)
     .def_property_readonly("size",FSize)
-    .def("gather",&PythonMPI::gather<double>)
-    .def("gather",&PythonMPI::gather<int>)
-    .def("gather", gather_list)
+    .def("gather", gather_int)
+    .def("gather", gather_double)
+    .def("gather", gather_list_int)
```
Comment (RiccardoRossi) [inline external_libraries/mpi_python/python_interface.cpp]: oh i see "gatherv" was not exposed. Btw, shouldn't we call this GatherV (as in the original call) rather than gather? (that is a question for @jcotela )

File: external_libraries/mpi_python/python_interface.cpp (hunk 2 above)
Comment (RiccardoRossi) [reply external_libraries/mpi_python/python_interface.cpp]: mmm, i just realized that this capability was not exposed before. Unfortunately this also implies that there is no test whatsoever for this :-(:-( ... which brings me to ask adding for at least a minimum coverage, nothing fancy but just to check it works

Comment (philbucher) [reply external_libraries/mpi_python/python_interface.cpp]: @RiccardoRossi no problem
Several things:
- This PR only refactors/cleans existing capabilities => I added tests already in a previous PR (#2619)
- The C-fct gives back an std::vector. Pybind automatically converts that to a regular python-list! You can do everything with it that you can do normally...
- This automatic conversion you don't like too much right? Not sure if we should change it though, seems like a "goodie" of pybind

File: external_libraries/mpi_python/python_interface.cpp (hunk 3 above)
Comment (philbucher) [reply external_libraries/mpi_python/python_interface.cpp]: yep this is GatherV

I can change it if wanted, but this will break existing code

**Comment by philbucher** [comment]
more comments @RiccardoRossi ?

Or do you want to wait until after your vacation?

File: external_libraries/mpi_python/python_interface.cpp (hunk 3 above)
Comment (philbucher) [reply external_libraries/mpi_python/python_interface.cpp]: I guess now would be the time to break things if I anyway refactor it

afaik only @mpentek and @msandre are using this, the impact could be limited

Comment (jcotela) [reply external_libraries/mpi_python/python_interface.cpp]: @RiccardoRossi I believe this function was added by @msandre. I think we can change it to gatherv for consistency, but I also think we can do it on a separate PR.

**Comment by RiccardoRossi** [comment]
i leave this to @jcotela. Don't need to wait for my holidays :-)

**Comment by philbucher** [comment]
after discussing with @msandre and @jcotela I changed the function-name from `gather` to `gatherv`, consistent with the underlying MPI-calls

This way I could also remove the function ptrs needed for exposing

**Review by jcotela — APPROVED** [review]

**Comment by philbucher** [comment]
I just checked, the version of gather that I renamed to gatherv was not being used in Master
